larpandoracontent/LArControlFlow/StitchingCosmicRayMergingTool.cc

```diff
@@ -590,19 +593,138 @@ void StitchingCosmicRayMergingTool::StitchPfos(const MasterAlgorithm *const pAlg
+    PfoVector::const_iterator iter = pfoVector.begin();
+    return *iter;
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+PfoVector StitchingCosmicRayMergingTool::SelectLongestStitch(PfoVector &pfoVector, const PfoToLArTPCMap &pfoToLArTPCMap) const
+{
+    PfoVector selectedPfos;
+    int totalHitsStitched(0);
+
+    for (const ParticleFlowObject *const pPfoToShift1 : pfoVector)
+    {
+        for (const ParticleFlowObject *const pPfoToShift2 : pfoVector)
+        {
+            if (pPfoToShift2 == pPfoToShift1)
+                continue;
+
+            const unsigned int twoDHitsPfo1(this->NumberOfTwoDHits(pPfoToShift1)), twoDHitsPfo2(this->NumberOfTwoDHits(pPfoToShift2));
+
+            PfoToLArTPCMap::const_iterator iter1(pfoToLArTPCMap.find(pPfoToShift1));
+            PfoToLArTPCMap::const_iterator iter2(pfoToLArTPCMap.find(pPfoToShift2));
+
+            if ((iter1 != pfoToLArTPCMap.end()) && (iter2 != pfoToLArTPCMap.end()))
+            {
+                const LArTPC *const pLArTPC1(iter1->second);
+                const LArTPC *const pLArTPC2(iter2->second);
+
+                if (((twoDHitsPfo1 + twoDHitsPfo2) > totalHitsStitched) && LArStitchingHelper::CanTPCsBeStitched(*pLArTPC1, *pLArTPC2))
+                {
+                    totalHitsStitched = twoDHitsPfo1 + twoDHitsPfo2;
+                }
+            }
+        }
+    }
+
+    // TODO - better way to save the pfos? a map pfo, total hits created in loop above?
+    for (const ParticleFlowObject *const pPfoToShift1 : pfoVector)
+    {
+        for (const ParticleFlowObject *const pPfoToShift2 : pfoVector)
+        {
+            if (pPfoToShift2 == pPfoToShift1)
+                continue;
+
+            const unsigned int twoDHitsPfo1(this->NumberOfTwoDHits(pPfoToShift1)), twoDHitsPfo2(this->NumberOfTwoDHits(pPfoToShift2));
+
+            if ((twoDHitsPfo1 + twoDHitsPfo2) == totalHitsStitched)
+            {
+                selectedPfos.push_back(pPfoToShift1);
+                selectedPfos.push_back(pPfoToShift2);
+                return selectedPfos;
+            }
+        }
+    }
+
+    //ATTN we shouldnt arrive here, but if we did, just resize to 2 (pfoVector is ordered by number of hits anyway)
+    pfoVector.resize(2);
+    return pfoVector;
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+//This could be replaced by LArPfoHelper Get2DClusters and then LArClusterHelper GetOrderedCaloHitList, or this below could be moved to LArPfoHelper
+unsigned int StitchingCosmicRayMergingTool::NumberOfTwoDHits(const ParticleFlowObject *pfo) const
+{
+    unsigned int totalHits(0);
+    for (ClusterList::const_iterator iter = pfo->GetClusterList().begin(), iterEnd = pfo->GetClusterList().end(); iter != iterEnd; ++iter)
```

Comment:
Old-style - now prefer range-based for loop, e.g. `for (const Cluster *const pCluster : pfo->GetClusterList())`